extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,342 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Driver for Sensirion SHT3x digital temperature and humity sensor
+ * connected to I2C
+ *
+ * Part of esp-open-rtos
+ */
+
+#include <string.h>
+
+#include "sht3x.h"
+
+#include "FreeRTOS.h"
+
+#include "espressif/esp_common.h"
+#include "espressif/sdk_private.h"
+
+#define SHT3x_STATUS_CMD               0xF32D
+#define SHT3x_CLEAR_STATUS_CMD         0x3041
+#define SHT3x_RESET_CMD                0x30A2
+#define SHT3x_FETCH_DATA_CMD           0xE000
+#define SHT3x_HEATER_OFF_CMD           0x3066
+
+uint16_t SHT3x_MEASURE_CMD[6][3] = { {0x2c06,0x2c0d,0x2c10},    // [SINGLE_SHOT][H,M,L]
+                                     {0x2032,0x2024,0x202f},    // [PERIODIC_05][H,M,L]
+                                     {0x2130,0x2126,0x212d},    // [PERIODIC_05][H,M,L]
+                                     {0x2236,0x2220,0x222b},    // [PERIODIC_05][H,M,L]
+                                     {0x2234,0x2322,0x2329},    // [PERIODIC_05][H,M,L]
+                                     {0x2737,0x2721,0x272a} };  // [PERIODIC_05][H,M,L]
+                                     
+int32_t SHT3x_MEASURE_DURATION[3] = {30,20,20};  // [High, Medium, Low]
+
+#ifdef SHT3x_DEBUG
+#define debug(s, f, ...) printf("%s %s: " s "\n", "SHT3x", f, ## __VA_ARGS__)
+#define debug_dev(s, f, d, ...) printf("%s %s: bus %d, addr %02x - " s "\n", "SHT3x", f, d->bus, d->addr, ## __VA_ARGS__)
+#else
+#define debug(s, f, ...)
+#define debug_dev(s, f, d, ...)
+#endif
+
+#define error(s, f, ...) printf("%s %s: " s "\n", "SHT3x", f, ## __VA_ARGS__)
+#define error_dev(s, f, d, ...) printf("%s %s: bus %d, addr %02x - " s "\n", "SHT3x", f, d->bus, d->addr, ## __VA_ARGS__)
+
+/** Forward declaration of function for internal use */
+
+static bool sht3x_send_command  (sht3x_sensor_t*, uint16_t);
+static bool sht3x_read_data     (sht3x_sensor_t*, uint8_t*,  uint32_t);
+static bool sht3x_compute_values(sht3x_sensor_t*, uint8_t*);
+static bool sht3x_reset         (sht3x_sensor_t*);
+static bool sht3x_get_status    (sht3x_sensor_t*, uint16_t*);
+
+/** ------------------------------------------------ */
+
+bool sht3x_init_driver()
+{
+    return true;
+}
+
+
+bool sht3x_init_sensor(sht3x_sensor_t* dev, uint8_t bus, uint8_t addr)
+{
+    if (!dev) return false;
+      
+    // init sensor data structure
+    dev->bus  = bus;
+    dev->addr = addr;
+    dev->mode = single_shot;
+    dev->repeatability = high;
+    dev->meas_started = false;
+    dev->meas_start_time = 0;
+    
+    dev->active = true;
```

Comment:
I have changed it as following: 

The function **_sht3x_init_sensor_** creates the data structure of type **_sht3x_sensor_t_** dynamically and returns the pointer to this data structure on success. In case of error, the function frees the data structure and returns NULL. That is, resources are only allocated when the sensor device is present.